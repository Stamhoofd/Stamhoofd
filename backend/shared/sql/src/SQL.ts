import { SQLExpression } from "./SQLExpression";
import { SQLSelect } from "./SQLSelect";
import { SQLColumnExpression, SQLTableExpression, SQLWildcardSelectExpression } from "./SQLExpressions";
import { SQLJoin, SQLJoinType } from "./SQLJoin";

class StaticSQL {
    wildcard(namespace?: string) {
        return new SQLWildcardSelectExpression(namespace)
    }

    column(namespace: string, column: string);
    column(column: string);
    column(namespaceOrColumn: string, column?: string) {
        if (column === undefined) {
            return new SQLColumnExpression(namespaceOrColumn)
        }
        return new SQLColumnExpression(namespaceOrColumn, column)
    }

    table(namespace: string, table: string);
    table(table: string);
    table(namespaceOrTable: string, table?: string) {
        if (table === undefined) {
            return new SQLTableExpression(namespaceOrTable)
        }
        return new SQLTableExpression(namespaceOrTable, table)
    }

    select(...columns: SQLExpression[]) {
        if (columns.length === 0) {
            return new SQLSelect(this.wildcard())
        }
        return new SQLSelect(...columns)
    }

    leftJoin(table: SQLExpression) {
        return new SQLJoin(SQLJoinType.Left, table)
    }

    rightJoin(table: SQLExpression) {
        return new SQLJoin(SQLJoinType.Right, table)
    }

    innerJoin(table: SQLExpression) {
        return new SQLJoin(SQLJoinType.Inner, table)
    }

    join(table: SQLExpression) {
        return new SQLJoin(SQLJoinType.Inner, table)
    }
}

export const SQL = new StaticSQL();