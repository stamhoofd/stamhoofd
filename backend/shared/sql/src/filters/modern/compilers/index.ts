export * from './less';
export * from './greater';
export * from './equals';
export * from './in';
