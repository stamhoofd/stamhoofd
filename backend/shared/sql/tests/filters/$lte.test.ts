import { baseModernSQLFilterCompilers, createColumnFilter, SQLModernValueType } from '../../src/filters/modern/SQLModernFilter';
import { SQL } from '../../src/SQL';
import { TableDefinition, test, testMatch } from '../utils';

describe('$lte', () => {
    /**
     * Tests that should be repeated for all filter types
     */
    describe('Common checks', () => {
        // todo
    });

    it('can filter on $lte', async () => {
        const filters = {
            ...baseModernSQLFilterCompilers,
            age: createColumnFilter({ expression: SQL.column('age'), type: SQLModernValueType.Number, nullable: false }),
        };

        await test({
            filter: {
                age: {
                    $lte: 18,
                },
            },
            filters,
            query: {
                query: '`default`.`age` <= ?',
                params: [18],
            },
        });
    });

    it('can invert $lte inside', async () => {
        const filters = {
            ...baseModernSQLFilterCompilers,
            age: createColumnFilter({ expression: SQL.column('age'), type: SQLModernValueType.Number, nullable: false }),
        };

        await test({
            filter: {
                age: {
                    $not: {
                        $lte: 18,
                    },
                },
            },
            filters,
            query: {
                query: '`default`.`age` > ?',
                params: [18],
            },
        });
    });

    it('can invert $lte outside', async () => {
        const filters = {
            ...baseModernSQLFilterCompilers,
            age: createColumnFilter({ expression: SQL.column('age'), type: SQLModernValueType.Number, nullable: false }),
        };

        await test({
            filter: {
                $not: {
                    age: {
                        $lte: 18,
                    },
                },
            },
            filters,
            query: {
                query: '`default`.`age` > ?',
                params: [18],
            },
        });
    });

    describe('MySQL Behaviour', () => {
        const tableDefinition: TableDefinition = {
            settings: {
                type: 'json',
                nullable: true,
            },
            name: {
                type: 'varchar',
                nullable: true,
            },
            age: {
                type: 'int',
                nullable: true,
            },
        };
        const filters = {
            ...baseModernSQLFilterCompilers,
            'name': createColumnFilter(
                { expression: SQL.column('name'), type: SQLModernValueType.String, nullable: true },
            ),
            'age': createColumnFilter(
                { expression: SQL.column('age'), type: SQLModernValueType.Number, nullable: true },
            ),
            'settings.name': createColumnFilter({ expression: SQL.jsonValue(SQL.column('settings'), '$.name'), type: SQLModernValueType.JSONString, nullable: true },
            ),
            'settings': createColumnFilter({ expression: SQL.column('settings'), type: SQLModernValueType.JSONObject, nullable: true }),
            'settings.enabled': createColumnFilter({ expression: SQL.jsonValue(SQL.column('settings'), '$.enabled'), type: SQLModernValueType.JSONBoolean, nullable: true },
            ),
            'settings.age': createColumnFilter({ expression: SQL.jsonValue(SQL.column('settings'), '$.age'), type: SQLModernValueType.JSONNumber, nullable: true },
            ),
        };

        it('Comparing string values is case insensitive', async () => {
            await testMatch({
                tableDefinition,
                filters,
                rows: [
                    {
                        name: 'alice',
                    },
                    {
                        name: 'Alice',
                    },
                    {
                        name: 'ALICE',
                    },
                ],
                doMatch: [
                    {
                        name: {
                            $lte: 'b',
                        },
                    },
                    {
                        name: {
                            $lte: 'alicea',
                        },
                    },
                    {
                        name: {
                            $lte: 'Alicea',
                        },
                    },
                    {
                        name: {
                            $lte: 'Alice',
                        },
                    },
                    {
                        name: {
                            $lte: 'aLice',
                        },
                    },
                    {
                        name: {
                            $lte: 'ALICE',
                        },
                    },
                    {
                        name: {
                            $lte: 'alice',
                        },
                    },
                ],
                doNotMatch: [
                    {
                        name: {
                            $lte: '',
                        },
                    },
                    {
                        name: {
                            $lte: 'a',
                        },
                    },
                    {
                        name: {
                            $lte: 'Alic',
                        },
                    },
                    {
                        name: {
                            $lte: 'alic',
                        },
                    },
                ],
            });
        });

        it('Comparing JSON string values is case sensitive', async () => {
            await testMatch({
                tableDefinition,
                filters,
                rows: [
                    {
                        settings: {
                            name: 'alice',
                        },
                    },
                    {
                        settings: {
                            name: 'Alice',
                        },
                    },
                    {
                        settings: {
                            name: 'ALICE',
                        },
                    },
                ],
                doMatch: [
                    {
                        'settings.name': {
                            $lte: 'alicea',
                        },
                    },
                    {
                        'settings.name': {
                            $lte: 'Alicea',
                        },
                    },
                    {
                        'settings.name': {
                            $lte: 'ALICEA',
                        },
                    },
                    {
                        'settings.name': {
                            $lte: 'Alice',
                        },
                    },
                    {
                        'settings.name': {
                            $lte: 'aLice',
                        },
                    },
                    {
                        'settings.name': {
                            $lte: 'ALICE',
                        },
                    },
                    {
                        'settings.name': {
                            $lte: 'alice',
                        },
                    },
                ],
                doNotMatch: [
                    {
                        'settings.name': {
                            $lte: '',
                        },
                    },
                    {
                        'settings.name': {
                            $lte: 'a',
                        },
                    },
                    {
                        'settings.name': {
                            $lte: 'Alic',
                        },
                    },
                    {
                        'settings.name': {
                            $lte: 'alic',
                        },
                    },
                ],
            });
        });

        it('Comparing numbers values is numberic', async () => {
            await testMatch({
                tableDefinition,
                filters,
                rows: [
                    {
                        age: 10,
                    },
                ],
                doNotMatch: [
                    {
                        age: {
                            $lte: 9,
                        },
                    },
                    {
                        age: {
                            $lte: 0,
                        },
                    },
                    {
                        age: {
                            $lte: -9,
                        },
                    },
                    {
                        age: {
                            $lte: 1,
                        },
                    },
                ],
                doMatch: [
                    {
                        age: {
                            $lte: 10,
                        },
                    },
                    {
                        age: {
                            $lte: 10.1,
                        },
                    },
                    {
                        age: {
                            $lte: 12,
                        },
                    },
                    {
                        age: {
                            $lte: 20,
                        },
                    },
                ],
            });
        });

        it('Comparing JSON numbers values is numberic', async () => {
            await testMatch({
                tableDefinition,
                filters,
                rows: [
                    {
                        settings: {
                            age: 10,
                        },
                    },
                ],
                doNotMatch: [
                    {
                        'settings.age': {
                            $lte: 9,
                        },
                    },
                    {
                        'settings.age': {
                            $lte: 0,
                        },
                    },
                    {
                        'settings.age': {
                            $lte: -9,
                        },
                    },
                    {
                        'settings.age': {
                            $lte: 1,
                        },
                    },
                ],
                doMatch: [
                    {
                        'settings.age': {
                            $lte: 10,
                        },
                    },
                    {
                        'settings.age': {
                            $lte: 10.1,
                        },
                    },
                    {
                        'settings.age': {
                            $lte: 12,
                        },
                    },
                    {
                        'settings.age': {
                            $lte: 20,
                        },
                    },
                ],
            });
        });

        it('NULL values are always the smallest for JSON numbers', async () => {
            await testMatch({
                tableDefinition,
                filters,
                rows: [
                    {
                        settings: {
                            age: null,
                        },
                    },
                    {
                        settings: null,
                    },
                    {
                        settings: {},
                    },
                ],
                doMatch: [
                    {
                        'settings.age': {
                            $lte: -1000,
                        },
                    },
                    {
                        'settings.age': {
                            $lte: 0,
                        },
                    },
                    {
                        'settings.age': {
                            $lte: null,
                        },
                    },
                    {
                        'settings.age': {
                            $lte: 1000000,
                        },
                    },
                ],
            });
        });

        it('NULL values are always the smallest for JSON strings', async () => {
            await testMatch({
                tableDefinition,
                filters,
                rows: [
                    {
                        settings: {
                            name: null,
                        },
                    },
                    {
                        settings: null,
                    },
                    {
                        settings: {},
                    },
                ],
                doMatch: [
                    {
                        'settings.name': {
                            $lte: 'aaaa',
                        },
                    },
                    {
                        'settings.name': {
                            $lte: '',
                        },
                    },
                    {
                        'settings.name': {
                            $lte: null,
                        },
                    },
                ],
            });
        });

        it('NULL values are always the smallest for strings', async () => {
            await testMatch({
                tableDefinition,
                filters,
                rows: [
                    {
                        name: null,
                    },
                ],
                doMatch: [
                    {
                        name: {
                            $lte: '',
                        },
                    },
                    {
                        name: {
                            $lte: 'aaaa',
                        },
                    },
                    {
                        name: {
                            $lte: null,
                        },
                    },
                ],
            });
        });
    });
});
