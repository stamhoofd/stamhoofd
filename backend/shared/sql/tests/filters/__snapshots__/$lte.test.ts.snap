// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`$lte MySQL Behaviour Comparing JSON numbers values is numberic: SQL Query for filter: {"settings.age":{"$lte":10.1}} 1`] = `
{
  "params": [
    10.1,
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE (JSON_EXTRACT(\`test_table\`.\`settings\`,"$.age") = CAST('null' AS JSON) OR JSON_EXTRACT(\`test_table\`.\`settings\`,"$.age") IS NULL) OR JSON_EXTRACT(\`test_table\`.\`settings\`,"$.age") <= ?",
}
`;

exports[`$lte MySQL Behaviour Comparing JSON numbers values is numberic: SQL Query for filter: {"settings.age":{"$lte":10}} 1`] = `
{
  "params": [
    10,
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE (JSON_EXTRACT(\`test_table\`.\`settings\`,"$.age") = CAST('null' AS JSON) OR JSON_EXTRACT(\`test_table\`.\`settings\`,"$.age") IS NULL) OR JSON_EXTRACT(\`test_table\`.\`settings\`,"$.age") <= ?",
}
`;

exports[`$lte MySQL Behaviour Comparing JSON numbers values is numberic: SQL Query for filter: {"settings.age":{"$lte":12}} 1`] = `
{
  "params": [
    12,
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE (JSON_EXTRACT(\`test_table\`.\`settings\`,"$.age") = CAST('null' AS JSON) OR JSON_EXTRACT(\`test_table\`.\`settings\`,"$.age") IS NULL) OR JSON_EXTRACT(\`test_table\`.\`settings\`,"$.age") <= ?",
}
`;

exports[`$lte MySQL Behaviour Comparing JSON numbers values is numberic: SQL Query for filter: {"settings.age":{"$lte":20}} 1`] = `
{
  "params": [
    20,
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE (JSON_EXTRACT(\`test_table\`.\`settings\`,"$.age") = CAST('null' AS JSON) OR JSON_EXTRACT(\`test_table\`.\`settings\`,"$.age") IS NULL) OR JSON_EXTRACT(\`test_table\`.\`settings\`,"$.age") <= ?",
}
`;

exports[`$lte MySQL Behaviour Comparing JSON numbers values is numberic: SQL Query for not match filter: {"settings.age":{"$lte":-9}} 1`] = `
{
  "params": [
    -9,
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE (JSON_EXTRACT(\`test_table\`.\`settings\`,"$.age") = CAST('null' AS JSON) OR JSON_EXTRACT(\`test_table\`.\`settings\`,"$.age") IS NULL) OR JSON_EXTRACT(\`test_table\`.\`settings\`,"$.age") <= ?",
}
`;

exports[`$lte MySQL Behaviour Comparing JSON numbers values is numberic: SQL Query for not match filter: {"settings.age":{"$lte":0}} 1`] = `
{
  "params": [
    0,
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE (JSON_EXTRACT(\`test_table\`.\`settings\`,"$.age") = CAST('null' AS JSON) OR JSON_EXTRACT(\`test_table\`.\`settings\`,"$.age") IS NULL) OR JSON_EXTRACT(\`test_table\`.\`settings\`,"$.age") <= ?",
}
`;

exports[`$lte MySQL Behaviour Comparing JSON numbers values is numberic: SQL Query for not match filter: {"settings.age":{"$lte":1}} 1`] = `
{
  "params": [
    1,
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE (JSON_EXTRACT(\`test_table\`.\`settings\`,"$.age") = CAST('null' AS JSON) OR JSON_EXTRACT(\`test_table\`.\`settings\`,"$.age") IS NULL) OR JSON_EXTRACT(\`test_table\`.\`settings\`,"$.age") <= ?",
}
`;

exports[`$lte MySQL Behaviour Comparing JSON numbers values is numberic: SQL Query for not match filter: {"settings.age":{"$lte":9}} 1`] = `
{
  "params": [
    9,
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE (JSON_EXTRACT(\`test_table\`.\`settings\`,"$.age") = CAST('null' AS JSON) OR JSON_EXTRACT(\`test_table\`.\`settings\`,"$.age") IS NULL) OR JSON_EXTRACT(\`test_table\`.\`settings\`,"$.age") <= ?",
}
`;

exports[`$lte MySQL Behaviour Comparing JSON string values is case sensitive: SQL Query for filter: {"settings.name":{"$lte":"ALICE"}} 1`] = `
{
  "params": [
    "alice",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE (JSON_EXTRACT(\`test_table\`.\`settings\`,"$.name") = CAST('null' AS JSON) OR JSON_EXTRACT(\`test_table\`.\`settings\`,"$.name") IS NULL) OR JSON_VALUE(\`test_table\`.\`settings\`,"$.name" RETURNING CHAR CHARACTER SET utf8mb4) <= ?",
}
`;

exports[`$lte MySQL Behaviour Comparing JSON string values is case sensitive: SQL Query for filter: {"settings.name":{"$lte":"ALICEA"}} 1`] = `
{
  "params": [
    "alicea",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE (JSON_EXTRACT(\`test_table\`.\`settings\`,"$.name") = CAST('null' AS JSON) OR JSON_EXTRACT(\`test_table\`.\`settings\`,"$.name") IS NULL) OR JSON_VALUE(\`test_table\`.\`settings\`,"$.name" RETURNING CHAR CHARACTER SET utf8mb4) <= ?",
}
`;

exports[`$lte MySQL Behaviour Comparing JSON string values is case sensitive: SQL Query for filter: {"settings.name":{"$lte":"Alice"}} 1`] = `
{
  "params": [
    "alice",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE (JSON_EXTRACT(\`test_table\`.\`settings\`,"$.name") = CAST('null' AS JSON) OR JSON_EXTRACT(\`test_table\`.\`settings\`,"$.name") IS NULL) OR JSON_VALUE(\`test_table\`.\`settings\`,"$.name" RETURNING CHAR CHARACTER SET utf8mb4) <= ?",
}
`;

exports[`$lte MySQL Behaviour Comparing JSON string values is case sensitive: SQL Query for filter: {"settings.name":{"$lte":"Alicea"}} 1`] = `
{
  "params": [
    "alicea",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE (JSON_EXTRACT(\`test_table\`.\`settings\`,"$.name") = CAST('null' AS JSON) OR JSON_EXTRACT(\`test_table\`.\`settings\`,"$.name") IS NULL) OR JSON_VALUE(\`test_table\`.\`settings\`,"$.name" RETURNING CHAR CHARACTER SET utf8mb4) <= ?",
}
`;

exports[`$lte MySQL Behaviour Comparing JSON string values is case sensitive: SQL Query for filter: {"settings.name":{"$lte":"aLice"}} 1`] = `
{
  "params": [
    "alice",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE (JSON_EXTRACT(\`test_table\`.\`settings\`,"$.name") = CAST('null' AS JSON) OR JSON_EXTRACT(\`test_table\`.\`settings\`,"$.name") IS NULL) OR JSON_VALUE(\`test_table\`.\`settings\`,"$.name" RETURNING CHAR CHARACTER SET utf8mb4) <= ?",
}
`;

exports[`$lte MySQL Behaviour Comparing JSON string values is case sensitive: SQL Query for filter: {"settings.name":{"$lte":"alice"}} 1`] = `
{
  "params": [
    "alice",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE (JSON_EXTRACT(\`test_table\`.\`settings\`,"$.name") = CAST('null' AS JSON) OR JSON_EXTRACT(\`test_table\`.\`settings\`,"$.name") IS NULL) OR JSON_VALUE(\`test_table\`.\`settings\`,"$.name" RETURNING CHAR CHARACTER SET utf8mb4) <= ?",
}
`;

exports[`$lte MySQL Behaviour Comparing JSON string values is case sensitive: SQL Query for filter: {"settings.name":{"$lte":"alicea"}} 1`] = `
{
  "params": [
    "alicea",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE (JSON_EXTRACT(\`test_table\`.\`settings\`,"$.name") = CAST('null' AS JSON) OR JSON_EXTRACT(\`test_table\`.\`settings\`,"$.name") IS NULL) OR JSON_VALUE(\`test_table\`.\`settings\`,"$.name" RETURNING CHAR CHARACTER SET utf8mb4) <= ?",
}
`;

exports[`$lte MySQL Behaviour Comparing JSON string values is case sensitive: SQL Query for not match filter: {"settings.name":{"$lte":""}} 1`] = `
{
  "params": [
    "",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE (JSON_EXTRACT(\`test_table\`.\`settings\`,"$.name") = CAST('null' AS JSON) OR JSON_EXTRACT(\`test_table\`.\`settings\`,"$.name") IS NULL) OR JSON_VALUE(\`test_table\`.\`settings\`,"$.name" RETURNING CHAR CHARACTER SET utf8mb4) <= ?",
}
`;

exports[`$lte MySQL Behaviour Comparing JSON string values is case sensitive: SQL Query for not match filter: {"settings.name":{"$lte":"Alic"}} 1`] = `
{
  "params": [
    "alic",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE (JSON_EXTRACT(\`test_table\`.\`settings\`,"$.name") = CAST('null' AS JSON) OR JSON_EXTRACT(\`test_table\`.\`settings\`,"$.name") IS NULL) OR JSON_VALUE(\`test_table\`.\`settings\`,"$.name" RETURNING CHAR CHARACTER SET utf8mb4) <= ?",
}
`;

exports[`$lte MySQL Behaviour Comparing JSON string values is case sensitive: SQL Query for not match filter: {"settings.name":{"$lte":"a"}} 1`] = `
{
  "params": [
    "a",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE (JSON_EXTRACT(\`test_table\`.\`settings\`,"$.name") = CAST('null' AS JSON) OR JSON_EXTRACT(\`test_table\`.\`settings\`,"$.name") IS NULL) OR JSON_VALUE(\`test_table\`.\`settings\`,"$.name" RETURNING CHAR CHARACTER SET utf8mb4) <= ?",
}
`;

exports[`$lte MySQL Behaviour Comparing JSON string values is case sensitive: SQL Query for not match filter: {"settings.name":{"$lte":"alic"}} 1`] = `
{
  "params": [
    "alic",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE (JSON_EXTRACT(\`test_table\`.\`settings\`,"$.name") = CAST('null' AS JSON) OR JSON_EXTRACT(\`test_table\`.\`settings\`,"$.name") IS NULL) OR JSON_VALUE(\`test_table\`.\`settings\`,"$.name" RETURNING CHAR CHARACTER SET utf8mb4) <= ?",
}
`;

exports[`$lte MySQL Behaviour Comparing numbers values is numberic: SQL Query for filter: {"age":{"$lte":10.1}} 1`] = `
{
  "params": [
    10.1,
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE \`test_table\`.\`age\` IS NULL OR \`test_table\`.\`age\` <= ?",
}
`;

exports[`$lte MySQL Behaviour Comparing numbers values is numberic: SQL Query for filter: {"age":{"$lte":10}} 1`] = `
{
  "params": [
    10,
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE \`test_table\`.\`age\` IS NULL OR \`test_table\`.\`age\` <= ?",
}
`;

exports[`$lte MySQL Behaviour Comparing numbers values is numberic: SQL Query for filter: {"age":{"$lte":12}} 1`] = `
{
  "params": [
    12,
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE \`test_table\`.\`age\` IS NULL OR \`test_table\`.\`age\` <= ?",
}
`;

exports[`$lte MySQL Behaviour Comparing numbers values is numberic: SQL Query for filter: {"age":{"$lte":20}} 1`] = `
{
  "params": [
    20,
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE \`test_table\`.\`age\` IS NULL OR \`test_table\`.\`age\` <= ?",
}
`;

exports[`$lte MySQL Behaviour Comparing numbers values is numberic: SQL Query for not match filter: {"age":{"$lte":-9}} 1`] = `
{
  "params": [
    -9,
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE \`test_table\`.\`age\` IS NULL OR \`test_table\`.\`age\` <= ?",
}
`;

exports[`$lte MySQL Behaviour Comparing numbers values is numberic: SQL Query for not match filter: {"age":{"$lte":0}} 1`] = `
{
  "params": [
    0,
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE \`test_table\`.\`age\` IS NULL OR \`test_table\`.\`age\` <= ?",
}
`;

exports[`$lte MySQL Behaviour Comparing numbers values is numberic: SQL Query for not match filter: {"age":{"$lte":1}} 1`] = `
{
  "params": [
    1,
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE \`test_table\`.\`age\` IS NULL OR \`test_table\`.\`age\` <= ?",
}
`;

exports[`$lte MySQL Behaviour Comparing numbers values is numberic: SQL Query for not match filter: {"age":{"$lte":9}} 1`] = `
{
  "params": [
    9,
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE \`test_table\`.\`age\` IS NULL OR \`test_table\`.\`age\` <= ?",
}
`;

exports[`$lte MySQL Behaviour Comparing string values is case insensitive: SQL Query for filter: {"name":{"$lte":"ALICE"}} 1`] = `
{
  "params": [
    "alice",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE \`test_table\`.\`name\` IS NULL OR \`test_table\`.\`name\` <= ?",
}
`;

exports[`$lte MySQL Behaviour Comparing string values is case insensitive: SQL Query for filter: {"name":{"$lte":"Alice"}} 1`] = `
{
  "params": [
    "alice",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE \`test_table\`.\`name\` IS NULL OR \`test_table\`.\`name\` <= ?",
}
`;

exports[`$lte MySQL Behaviour Comparing string values is case insensitive: SQL Query for filter: {"name":{"$lte":"Alicea"}} 1`] = `
{
  "params": [
    "alicea",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE \`test_table\`.\`name\` IS NULL OR \`test_table\`.\`name\` <= ?",
}
`;

exports[`$lte MySQL Behaviour Comparing string values is case insensitive: SQL Query for filter: {"name":{"$lte":"aLice"}} 1`] = `
{
  "params": [
    "alice",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE \`test_table\`.\`name\` IS NULL OR \`test_table\`.\`name\` <= ?",
}
`;

exports[`$lte MySQL Behaviour Comparing string values is case insensitive: SQL Query for filter: {"name":{"$lte":"alice"}} 1`] = `
{
  "params": [
    "alice",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE \`test_table\`.\`name\` IS NULL OR \`test_table\`.\`name\` <= ?",
}
`;

exports[`$lte MySQL Behaviour Comparing string values is case insensitive: SQL Query for filter: {"name":{"$lte":"alicea"}} 1`] = `
{
  "params": [
    "alicea",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE \`test_table\`.\`name\` IS NULL OR \`test_table\`.\`name\` <= ?",
}
`;

exports[`$lte MySQL Behaviour Comparing string values is case insensitive: SQL Query for filter: {"name":{"$lte":"b"}} 1`] = `
{
  "params": [
    "b",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE \`test_table\`.\`name\` IS NULL OR \`test_table\`.\`name\` <= ?",
}
`;

exports[`$lte MySQL Behaviour Comparing string values is case insensitive: SQL Query for not match filter: {"name":{"$lte":""}} 1`] = `
{
  "params": [
    "",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE \`test_table\`.\`name\` IS NULL OR \`test_table\`.\`name\` <= ?",
}
`;

exports[`$lte MySQL Behaviour Comparing string values is case insensitive: SQL Query for not match filter: {"name":{"$lte":"Alic"}} 1`] = `
{
  "params": [
    "alic",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE \`test_table\`.\`name\` IS NULL OR \`test_table\`.\`name\` <= ?",
}
`;

exports[`$lte MySQL Behaviour Comparing string values is case insensitive: SQL Query for not match filter: {"name":{"$lte":"a"}} 1`] = `
{
  "params": [
    "a",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE \`test_table\`.\`name\` IS NULL OR \`test_table\`.\`name\` <= ?",
}
`;

exports[`$lte MySQL Behaviour Comparing string values is case insensitive: SQL Query for not match filter: {"name":{"$lte":"alic"}} 1`] = `
{
  "params": [
    "alic",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE \`test_table\`.\`name\` IS NULL OR \`test_table\`.\`name\` <= ?",
}
`;

exports[`$lte MySQL Behaviour NULL values are always the smallest for JSON numbers: SQL Query for filter: {"settings.age":{"$lte":-1000}} 1`] = `
{
  "params": [
    -1000,
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE (JSON_EXTRACT(\`test_table\`.\`settings\`,"$.age") = CAST('null' AS JSON) OR JSON_EXTRACT(\`test_table\`.\`settings\`,"$.age") IS NULL) OR JSON_EXTRACT(\`test_table\`.\`settings\`,"$.age") <= ?",
}
`;

exports[`$lte MySQL Behaviour NULL values are always the smallest for JSON numbers: SQL Query for filter: {"settings.age":{"$lte":0}} 1`] = `
{
  "params": [
    0,
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE (JSON_EXTRACT(\`test_table\`.\`settings\`,"$.age") = CAST('null' AS JSON) OR JSON_EXTRACT(\`test_table\`.\`settings\`,"$.age") IS NULL) OR JSON_EXTRACT(\`test_table\`.\`settings\`,"$.age") <= ?",
}
`;

exports[`$lte MySQL Behaviour NULL values are always the smallest for JSON numbers: SQL Query for filter: {"settings.age":{"$lte":1000000}} 1`] = `
{
  "params": [
    1000000,
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE (JSON_EXTRACT(\`test_table\`.\`settings\`,"$.age") = CAST('null' AS JSON) OR JSON_EXTRACT(\`test_table\`.\`settings\`,"$.age") IS NULL) OR JSON_EXTRACT(\`test_table\`.\`settings\`,"$.age") <= ?",
}
`;

exports[`$lte MySQL Behaviour NULL values are always the smallest for JSON numbers: SQL Query for filter: {"settings.age":{"$lte":null}} 1`] = `
{
  "params": [],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE JSON_EXTRACT(\`test_table\`.\`settings\`,"$.age") = CAST('null' AS JSON) OR JSON_EXTRACT(\`test_table\`.\`settings\`,"$.age") IS NULL",
}
`;

exports[`$lte MySQL Behaviour NULL values are always the smallest for JSON strings: SQL Query for filter: {"settings.name":{"$lte":""}} 1`] = `
{
  "params": [
    "",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE (JSON_EXTRACT(\`test_table\`.\`settings\`,"$.name") = CAST('null' AS JSON) OR JSON_EXTRACT(\`test_table\`.\`settings\`,"$.name") IS NULL) OR JSON_VALUE(\`test_table\`.\`settings\`,"$.name" RETURNING CHAR CHARACTER SET utf8mb4) <= ?",
}
`;

exports[`$lte MySQL Behaviour NULL values are always the smallest for JSON strings: SQL Query for filter: {"settings.name":{"$lte":"aaaa"}} 1`] = `
{
  "params": [
    "aaaa",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE (JSON_EXTRACT(\`test_table\`.\`settings\`,"$.name") = CAST('null' AS JSON) OR JSON_EXTRACT(\`test_table\`.\`settings\`,"$.name") IS NULL) OR JSON_VALUE(\`test_table\`.\`settings\`,"$.name" RETURNING CHAR CHARACTER SET utf8mb4) <= ?",
}
`;

exports[`$lte MySQL Behaviour NULL values are always the smallest for JSON strings: SQL Query for filter: {"settings.name":{"$lte":null}} 1`] = `
{
  "params": [],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE JSON_EXTRACT(\`test_table\`.\`settings\`,"$.name") = CAST('null' AS JSON) OR JSON_EXTRACT(\`test_table\`.\`settings\`,"$.name") IS NULL",
}
`;

exports[`$lte MySQL Behaviour NULL values are always the smallest for strings: SQL Query for filter: {"name":{"$lte":""}} 1`] = `
{
  "params": [
    "",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE \`test_table\`.\`name\` IS NULL OR \`test_table\`.\`name\` <= ?",
}
`;

exports[`$lte MySQL Behaviour NULL values are always the smallest for strings: SQL Query for filter: {"name":{"$lte":"aaaa"}} 1`] = `
{
  "params": [
    "aaaa",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE \`test_table\`.\`name\` IS NULL OR \`test_table\`.\`name\` <= ?",
}
`;

exports[`$lte MySQL Behaviour NULL values are always the smallest for strings: SQL Query for filter: {"name":{"$lte":null}} 1`] = `
{
  "params": [],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE \`test_table\`.\`name\` IS NULL",
}
`;
